### ISSUE TITLE
[v5] Avoid ad-hoc `State` creation

### ISSUE THREAD

**Issue by davidkpiano** [statelyai/xstate#2576]
I like the overall idea of this PR - this seems like a very useful picture. However, I don't quite like how this fits in the existing API. This is by no means a blocker for this PR but I would appreciate it if we could reconsider how this works in the v5. 

I'm OK with `State` having a link to a machine to use it for stuff like that but that IMHO means that the `State` should only be created by machines - users should not be responsible for creating it. The `machine` option accepted by the constructor doesn't seem like an option that can change the behavior of this object (its method) in such a profound way. It is its core responsibility.

I understand that people might need to serialize and deserialize states but then the deserialization process should result in a different kind of object that would be fed to the machine (its `resolveState` or whatever) so the final `State` could be created.

In other words - this change IMHO warrants hiding the `State`'s constructor, in the future, as the private API. It would still be public in the sense that people would interact with it when they receive it from callbacks, state-related properties on a machine/intepreter etc.

_Originally posted by @Andarist in https://github.com/statelyai/xstate/pull/2546#discussion_r696688811_

**Comment by Andarist** [comment]
this was largely addressed by https://github.com/statelyai/xstate/pull/4440 and some of the other changes that we introduced before it
